Sync/BatchingClient.swift

```diff
@@ -0,0 +1,245 @@
+    }
+
+    mutating func add(payload: Payload) {
+        count += payload.count
+        bytes += payload.bytes
+    }
+}
+
+
+// MARK: Batching Client
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private(set) var ifUnmodifiedSince: Timestamp?
+
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private var currentBatch: Batch?
+    private var currentPayload: Payload
+
+    private var onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.currentPayload = Payload(config: config)
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func endBatch() -> Success {
+        func uploadPayload() -> Success {
+            let lines = self.currentPayload.records.map { $0.payload }
+            let batchStartParam = NSURLQueryItem(name: "batch", value: "true")
+            let commitParam = NSURLQueryItem(name: "commit", value: "true")
+            return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: [batchStartParam, commitParam])
+                >>== effect(moveForward) >>> succeed
+        }
+
+        // Send up as a single upload if we're not batching.
+        guard let batch = self.currentBatch else {
+            return self.currentPayload.isEmpty ? succeed() : uploadPayload()
+        }
+
+        // If we have a batch and both it and the payload are empty, don't need to do anything
+        if batch.isEmpty && self.currentPayload.isEmpty {
+            return succeed()
+        }
+
+        if batch.isFull {
+            return commit(batch) >>> uploadPayload
+        } else {
+            return commit(batch) >>> succeed
+        }
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        if records.isEmpty {
+            return succeed()
+        }
+
+        do {
+            var generator = records.generate()
+            while let record = generator.next() {
+                let serializedRecord = try serialize(record)
+                if let uploadOp = accumulateRecord(serializedRecord) {
+                    // Since we couldn't add this record, throw it onto the rest and try adding everything again.
+                    let restOfRecords = generator.map { $0 }
+                    return uploadOp >>> { self.addRecords([record] + restOfRecords) }
```

Comment:
Rather than consuming the entire generator, concatenating a whole new array, and starting again, do something like this:

```
if let uploadOp = accumulateRecord(serializedRecord) {
    // An upload started. Chain off it to add the triggering record, and pick up
    // where we left off.
    return uploadOp >>> {
        let deferred = accumulateRecord(serializedRecord) ?? succeed()  // Might need type fiddling.
        return deferred >>> self.addRecords(generator)     // Add the generator variant.
    }
```